### PR TITLE
[Fix] `stringify`: actually fix cyclic references

### LINKED ISSUE #403: qs.stringify Uncaught RangeError: Cyclic object value
var qs = require('qs');
var a = ['a'];
qs.stringify({x:a,y:a});

-> Uncaught RangeError: Cyclic object value

### PR DESCRIPTION
**Pull request by liaokunhua**
With all the testing, I think this is the ultimate solution for circular references.

Fixes #403.

### CODE REVIEW

File: lib/stringify.js [hunk 1]
```diff
@@ -56,6 +56,8 @@ var isNonNullishPrimitive = function isNonNullishPrimitive(v) {
         || typeof v === 'bigint';
 };
 
+var scSym = Symbol('SideChannel');
```
Comment (ljharb) [inline lib/stringify.js:59]: We support many engines that do not have symbols.

File: lib/stringify.js [hunk 2]
```diff
@@ -75,8 +77,23 @@ var stringify = function stringify(
...
-    if (sideChannel.has(object)) {
-        throw new RangeError('Cyclic object value');
+    var tmpSc = sideChannel;
+    var step = 0;
+    var findFlag = false;
+    while ((tmpSc = tmpSc.get(scSym)) !== undefined && !findFlag) {
+        // Where object last appeared in the ref tree
+        var pos = tmpSc.get(object);
+        step += 1;
+        if (pos != undefined) {
```
Comment (ljharb) [inline lib/stringify.js:87]: This should use !==, or != null; != should never be used to compare to anything besides null.

**Review by ljharb — CHANGES_REQUESTED** [review]
Can you provide a regression test that this fixes?

**Comment by liaokunhua** [comment]
> Can you provide a regression test that this fixes?

Sorry bro, this is a bit out of my league, I haven't thought of a special object that would prove i didn't fix this bug. 

So, let's talk about how I found this bug.
I started learning front-end and github a few days ago, starting with a simple project.
I noticed that the stringify.js file in the test report had a line of code that was not covered.

=> `throw new RangeError('Cyclic object value');`

I did some tests, the result is that non-circular duplicated references can work, but circular dependencies result in "RangeError: Maximum call stack size exceeded", not "RangeError: Cyclic object value".

I think the exception thrown in the use case test of the circular dependency happens to be RangeError, but the exception message is not checked.

File: lib/stringify.js (hunk 1 above)
Comment (liaokunhua) [reply lib/stringify.js:59]: emmm....it was my poor consideration, I meant to define a unique symbol to identify the side-channel, and Symbol is good idea.

File: lib/stringify.js (hunk 2 above)
Comment (liaokunhua) [reply lib/stringify.js:87]: i get it, thanks for the guidance.

Comment (liaokunhua) [reply lib/stringify.js:87]: > 

i get it, thanks for the guidance.

**Comment by ljharb** [comment]
Surely you provided some input that caused that error?

File: lib/stringify.js [hunk 3]
```diff
@@ -56,7 +56,7 @@ var isNonNullishPrimitive = function isNonNullishPrimitive(v) {
         || typeof v === 'bigint';
 };
 
-var scSym = Symbol('SideChannel');
+var scSym = Object.create(null);
```
Comment (ljharb) [inline lib/stringify.js:59]: Unfortunately Object.create also can’t be relied upon.
```suggestion
var sentinel = {};
```

**Review by ljharb — CHANGES_REQUESTED** [review]

**Comment by liaokunhua** [comment]
> Surely you provided some input that caused that error?

Right, an example of a circular reference is constructed, which is inside the use case test.
It is easy to reproduce.

Based on the original code for testing.
Step1: 
modify line 79 in the stringify.js file, `throw new RangeError('Cyclic object value');` to `throw new Error('Cyclic object value');`.

Step2:
Create a new test and select an example of a circular reference in the stringify.js file inside the test folder and make the changes.
```
t.test("does not crash when parsing circular references", function(st) {
        var circular = { a: 'value' };
        circular.a = circular;
        try {
            qs.stringify(circular);
        } catch(e) {
            console.dir(e)
        }
        st['throws'](
            function () { qs.stringify(circular); },
            RangeError,
            'Cyclic values throw'
        );
        st.end();
})
```

Step3:
Run test and get output:
```
TAP version 13
# stringify()
# does not crash when parsing circular references
RangeError: Maximum call stack size exceeded
// ....omitted stack info
ok 1 Cyclic values throw

1..1
# tests 1
# pass  1

# ok
```

I was dumbfounded!! not the expected "Error:Cyclic object value".

Environment information:
system: ubuntu 20.04
node: v14.16.1
npm: 6.14.12

**Comment by ljharb** [comment]
ah, i see - so changing that test to check the error message - so we know it's not a stack overflow - would reproduce the error that this PR fixes?

if so, could you make that change?

**Comment by liaokunhua** [comment]
> ah, i see - so changing that test to check the error message - so we know it's not a stack overflow - would reproduce the error that this PR fixes?
> 
> if so, could you make that change?

emmm...kinda wish it was me, but i don't think i have it in me yet.😅

File: lib/stringify.js (hunk 3 above)
Comment (liaokunhua) [reply lib/stringify.js]: > 

Submitted new code. for this question, i didn't understand why it was unreliable. two objects created by `Object.create`, return value is false when judged by ===.

Comment (liaokunhua) [reply lib/stringify.js]: I see, polyfill is provided in MDN and it doesn't support null as the first parameter, is that the reason?

Comment (ljharb) [reply lib/stringify.js]: It's impossible to polyfill null objects.

We support down to IE 6, which doesn't have Object.create or `__proto__`.

**Comment by ljharb** [comment]
I rebased this PR, and updated the test cases in the default branch that were asserting a RangeError, but not the right one, and added another test case that confirms a fix for #403.

File: lib/stringify.js [hunk 4]
```diff
@@ -75,8 +77,23 @@ var stringify = function stringify(
 ) {
     var obj = object;
 
-    if (sideChannel.has(object)) {
-        throw new RangeError('Cyclic object value');
+    var tmpSc = sideChannel;
+    var step = 0;
+    var findFlag = false;
+    while ((tmpSc = tmpSc.get(sentinel)) !== undefined && !findFlag) {
```
Comment (ljharb) [inline lib/stringify.js:83]: is there any way to do this without a loop, or at least without doing an assignment inside the conditional expression?

**Review by ljharb — APPROVED** [review]
actually i'll merge this as-is, to fix the bug; but i'd love it if there was a followup to clean up the loop.

File: lib/stringify.js (hunk 4 above)
Comment (liaokunhua) [reply lib/stringify.js:83]: no other graceful way come to mind at the moment, and if I think of any later, I will submit another PR.

**Pull request merged**